src/prefect/utilities/executors.py

```diff
@@ -377,3 +393,41 @@ def prepare_upstream_states_for_mapping(
             break
 
     return map_upstream_states
+
+
+def _build_flattened_state(state: "State", index: int) -> "State":
+    """Helper function for `flatten_upstream_state`"""
+    new_state = copy.copy(state)
+    new_state.result = state._result.from_value(  # type: ignore
+        state.result[index]
+    )
+    return new_state
+
+
+def flatten_upstream_state(upstream_state: "State") -> "State":
+    new_state = copy.copy(upstream_state)
+    try:
+        flattened_result = [y for x in upstream_state.result for y in x]
```

Comment:
this basically duplicates the data. For a heavy mapped task (say, with 1000 tasks), that could be a problem, couldn't it? I'm not sure the rest of the code supports it, but would it work to change this to a generator instead?

Reply:
Generators are generally incompatible with serialization between tasks, so this would require a tight coupling between this feature and iteration within the task. I would suggest that we look at that for a future enhancements - perhaps tasks that are reducing over a mapped upstream can opt to process the upstream input as a sequence rather than a full in-memory reduction.